cou/utils/app_utils.py

```diff
@@ -24,17 +25,24 @@
 logger = logging.getLogger(__name__)
 
 
-async def upgrade_packages(units: Iterable[str], model: COUModel) -> None:
+async def upgrade_packages(
+    units: Iterable[str], model: COUModel, packages_to_hold: Optional[list] = None
+) -> None:
     """Run package updates and upgrades on each unit of an Application.
 
     :param units: The list of unit names where the package upgrade runs on.
     :type units: Iterable[str]
     :param model: COUModel object
     :type model: COUModel
+    :param packages_to_hold: A list of packages to put on hold during package upgrade.
+    :type packages_to_hold: Optional[list]
     :raises RunUpgradeError: When an upgrade fails.
     """
     dpkg_opts = "-o Dpkg::Options::=--force-confnew -o Dpkg::Options::=--force-confdef"
     command = f"apt-get update && apt-get dist-upgrade {dpkg_opts} -y && apt-get autoremove -y"
+    if packages_to_hold:
+        packages = " ".join(packages_to_hold)
+        command = f"sudo apt-mark hold {packages} && {command} && sudo apt-mark unhold {packages}"
```

Comment:
What if the upgrade fails, but the hold does not? You could consider splitting the three actions (hold/upgrade/unhold) to more easily respond to failures, or at least turn the last `&&` into an or to let the unhold command run in (almost) any case

Reply:
Thanks for the suggestion.

Running commands 3 times would not be ideal because we would need to establish connection to a unit multiple times. Since running `unhold` on non-hold package is safe, I chose to implement the second option.

Reply:
This is really good catch and I think that second approach is better, since connecting to unit multiple times can provides simply more room to failure.